franka_control_node incorrectly handles controller starting

1. Trying to start a `ros_control` controller when the robot is not in automatic mode (external activation device pushed down), `franka_control_node` correctly rejects the operation: 
`[ERROR] [ros.franka_control]: libfranka: Move command rejected: command not possible in the current mode!`
But: the controller is reported as started, e.g. in rqt's controller manager plugin.
2. If the robot is switched back to automatic mode (external activation device pulled up), starting a controller is still not possible:
    - `franka_control_node` doesn't report an error (anymore)
    - the controller is reported to be active in rqt, but the controller isn't actually running!
3. To enable starting a controller, one needs to re-launch `franka_control_node`, which is rather annoying if it is part of a larger roslaunch file.

Hi @rhaschke do you have any update on this? We have the same problem.

We do the following:

- Launch the franka_control.launch from the franka_control package.
- Spawn the position_joint_trajectory_controller using the controller_manager.

It appears that the position_joint_trajectory_controller starts (as the appropriate topics appear), but we also get that error: `[ERROR] [ros.franka_control]: libfranka: Move command rejected: command not possible in the current mode!`.

We don't know if it's related, but when using MoveIt! to move the arm, MoveIt reports: "Controller 'position_joint_trajectory_controller' failed with error GOAL_TOLERANCE_VIOLATED:" even though MoveIt! finds a trajectory.

The same MoveIt! configuration works out-of-the-box when using Gazebo, so I doubt that it's a moveit issue.

--- a/franka_control/src/franka_control_node.cpp
+++ b/franka_control/src/franka_control_node.cpp
@@ -139,6 +139,7 @@ int main(int argc, char** argv) {
       }
     }
 
+    ROS_INFO_THROTTLE(1, "franka_control: controller activated");
     if (franka_control.connected()) {
       try {
         // Run control loop. Will exit if the controller is switched.
@@ -160,7 +161,7 @@ int main(int argc, char** argv) {
         has_error = true;
       }
     }
-    ROS_INFO_THROTTLE(1, "franka_control, main loop");
+    ROS_INFO_THROTTLE(1, "franka_control: main loop");
   }
 
   return 0;

--- a/franka_control/src/franka_control_node.cpp
+++ b/franka_control/src/franka_control_node.cpp
@@ -124,6 +124,10 @@ int main(int argc, char** argv) {
           control_manager.update(now, now - last_time);
           franka_control.checkJointLimits();
           last_time = now;
+
+          if (has_error && franka_control.robotMode() == franka::RobotMode::kIdle) {
+            has_error = false;
+          }
         } catch (const std::logic_error& e) {
         }
       } else {

--- a/franka_hw/include/franka_hw/franka_hw.h
+++ b/franka_hw/include/franka_hw/franka_hw.h
@@ -295,6 +295,8 @@ class FrankaHW : public hardware_interface::RobotHW {
                                                     const ros::NodeHandle& robot_hw_nh,
                                                     const std::vector<double>& defaults);
 
+  franka::RobotMode robotMode() const { return robot_state_ros_.robot_mode; }
+
  protected:
   /**
    * Checks whether an array of doubles contains NaN values.

